compiler/rustc_target/src/abi/call/x86.rs

```diff
@@ -53,11 +53,38 @@ where
         if arg.is_ignore() {
             continue;
         }
-        if arg.layout.is_aggregate() {
-            arg.make_indirect_byval();
-        } else {
+        if !arg.layout.is_aggregate() {
             arg.extend_integer_width_to(32);
+            continue;
         }
+
+        // We need to compute the alignment of the `byval` argument. The rules can be found in
+        // `X86_32ABIInfo::getTypeStackAlignInBytes` in Clang's `TargetInfo.cpp`. Summarized here,
+        // they are:
+        //
+        // 1. If the natural alignment of the type is less than or equal to 4, the alignment is 4.
+        //
+        // 2. Otherwise, on Linux, the alignment of any vector type is the natural alignment.
+        // (This doesn't matter here because we ensure we have an aggregate with the check above.)
```

Comment:
This will become a problem is https://github.com/rust-lang/rust/issues/63068 is ever implemented. TBH I think this code should just omit vector handling entirely for now, rather than providing some very dubious, untested implementation for Darwin only.

Reply:
That sounds reasonable, but what do you want it to do instead?

My initial thought was to error here for any vector ABI or vector-containing aggregate.
I think the former (error on vector ABI) is not really feasible, because the `simd_ffi` feature gate is at least partially implemented, and there is [a test](https://github.com/rust-lang/rust/blob/8d18c32b61476ed16dd15074e71be3970368d6d7/tests/run-make/simd-ffi/simd.rs#L15) that exercises SIMD types in FFI.
The latter (error on aggregate containing vector) is feasible, although erroring also isn't great, since it doesn't seem like there's a good way to report a nice error here, only an ICE, but maybe that's fine...

Or did you mean just remove the vector case i.e. have it always use align 4?

Since I've already written the code to detect vector-containing aggregates, I kinda want to just implement the full correct logic...

Reply:
> I kinda want to just implement the full correct logic...

I have done so (with a test) in the latest commit. Let me know if you'd prefer something else